KAFKA-2593 Key value stores can use specified serializers and deserializers

rhauch opened the pull request:
Add support for the key value stores to use specified serializers and deserializers (aka, "serdes"). Prior to this change, the stores were limited to only the default serdes specified in the topology's configuration and exposed to the processors via the ProcessorContext.

Now, using InMemoryKeyValueStore and RocksDBKeyValueStore are similar: both are parameterized on the key and value types, and both have similar multiple static factory methods. The static factory methods either take explicit key and value serdes, take key and value class types so the serdes can be inferred (only for the built-in serdes for string, integer, long, and byte array types), or use the default serdes on the ProcessorContext.

rhauch committed:
KAFKA-2593 Key value stores can use custom serializers and deserializers

Add support for the key value stores to use specified serializers and deserializers (aka, "serdes"). Prior to this change, the stores were limited to only the default serdes specified in the topology's configuration and exposed to the processors via the ProcessorContext.

Now, using InMemoryKeyValueStore and RocksDBKeyValueStore are similar: both are parameterized on the key and value types, and both have similar multiple static factory methods. The static factory methods either take explicit key and value serdes, take key and value class types so the serdes can be inferred (only for the built-in serdes for string, integer, long, and byte array types), or use the default serdes on the ProcessorContext.

rhauch commented:
I updated the PR to correct the checkstyle errors.

guozhangwang commented:
LGTM overall, one general comment: could we put various create() functions of different store classes into a single sharable support / factory class, otherwise future users would need to create those functions for each store?

rhauch commented:
@guozhangwang, sure, it would be possible to put all of the create methods in a single place. However, I've never been a fan of centralized factory methods because it makes other implementations _second-class citizens_, since _they_ can't be created using the factory. (Yes, we could use Java's `ServiceLoader` mechanism, but that's IMO that's overkill and still doesn't allow customization of create methods in other KeyValueStore implementations.)

Now, I do agree that the 2 existing implementations (or 3 including #256) all have similar factory methods with identical signatures, and that might suggest value to a common way to build them. But perhaps a builder (rather than a factory class) might be more usable by processors and reusable by other `KeyValueStore` implementations. For example:

```
InMemoryStore<String,Integer> store = InMemoryStore.create(name,context)
                                                   .withKeySerializer(new StringSerializer())
                                                   .withKeyDeserializer(new StringDeserializer())
                                                   .withValueSerializer(new IntegerSerializer())
                                                   .withValueDeserializer(new IntegerDeserializer())
                                                   .withSystemTime()
                                                   .build();
```

It would be possible to create a base interface and impl for this builder in `KeyValueStore` that subclasses could reuse _and extend_ with their own methods. The builder could have overloaded methods, too, without resulting in combinatorial explosion of separate factory methods. For example, the builder could have both `withSystemTime()` and `withTime(Time)`, and the last one used wins. And it could have multiple `withKeySerializer` methods that take different things, such as a serializer instance, the serializer class (to automatically create the instance using a no-arg constructor), or the key class (for built-in serdes only). IDEs also make builders pretty convenient to use.

This will expand the Kafka Streams code a bit, but IMO it is worth it because it will make it even easier (albeit a bit more verbose) to create `KeyValueStore` instances in processors.

I'll prototype something up, but should I do that as part of this PR/issue, or under a separate "Simplified approach to create KeyValueStores" issue?

rhauch commented:
@guozhangwang, it may be difficult or unattractive to use a builder with the generic key and value types, unless somehow they're specified in the `create` method or by explicitly typing:

```
InMemoryKeyValueStore<String, Integer> store = InMemoryKeyValueStore.<String, Integer>create("something", context)
                                               .withKeySerializer(new StringSerializer())
                                               .withKeyDeserializer(new StringDeserializer())
                                               .withValueSerializer(new IntegerSerializer())
                                               .withValueDeserializer(new IntegerDeserializer())
                                               .withSystemTime()
                                               .build();
```

The good news is that the above code forces the IDE to type-check the `with...` methods, but it's not pretty. I'll keep trying.

rhauch committed:
KAFKA-2593: MeteredKeyValueStore and others no longer cast to ProcessorContextImpl

The cast was used to access ProcessorContextImpl.recordCollector(), and the cast prevent using MeteredKeyValueStore subclasses, SinkNode, and SlidingWindow objects in tests where something other than ProcessorContextImpl was used. This change introduces a RecordCollector.Supplier interface to define this `recordCollector()` method, and changed ProcessorContextImpl and MockProcessorContext to both implement this interface. Now, MeteredKeyValueStore, SinkNode, and SlidingWindow can cast to the interface to access the record collector, making it possible to use them inside unit tests.

rhauch committed:
KAFKA-2593: Corrected RocksDBKeyValueStore to properly create directory if missing

rhauch committed:
KAFKA-2593 Added unit tests for the key value stores

Added unit tests for the existing InMemoryKeyValueStore and RocksDBKeyValueStore implementations. A new KeyValueStoreTestDriver class does most of the work and simplifies the tests.

rhauch committed:
KAFKA-2593 Correct how Kafka Streams tests create and use local state directory

The ProcessorTopologyTest and AbstractKeyValueStoreTest test classes have methods created a shared directory used by all their tests to store local state. When the computer has multiple processors, the build system executes test in parallel, and they interfere with other tests using the same directory. Solve the problem by adding a utility class to create temporary directories and automatically clean them up.

rhauch commented:
@guozhangwang, there doesn't appear to be a decent way to use a builder while returning a parameterized type. The example mentioned above is pretty verbose, and IMO the static factory methods on each KeyValueStore implementation class is the most pragmatic way to create instances. It's not ideal, because as you pointed out each class has static methods with identical (or at least nearly identical) signatures. OTOH, I'm not really sure how this might be simplified another way.

BTW, I'd like to pull some of the commits from #256 into this PR, which is really where they belong. Specifically, all but the one commit that creates the `InMemoryLRUCacheStore` and corresponding test class. I've already moved them over on a local branch and ran a build, so if you're okay with it I'll push the _additional_ commits to this PR and also update #256.

rhauch commented:
@guozhangwang, it occurred to me that your goal might actually be to make the implementations of `KeyValueStore` largely transparent to users, and instead to emphasize that stores can either keep all entries in-memory (and is thus largely limited by available heap) while the other holds entries outside of the heap and this is largely unlimited. Is this accurate?

If this is true, and since builders are relatively difficult to implement for parameterized stores, then I agree that a factory class might be the best way forward. Do you have time to list what you might have been thinking of for an API? Perhaps given some store defined in a `Processor`:

```
private KeyValueStore<String, Integer> myStore;
```

the store could be created in `init(...)` with something like:

```
myStore = Stores.inMemory()
                .create(name,context,...);
```

or with limited memory (see #256):

```
myStore = Stores.inMemory()
                .maxCount(4096)
                .create(name,context,...);
```

or to use the database:

```
myStore = Stores.offHeap()
                .create(name,context,...);
```

What do you think? This is somewhat like a builder, but all of the parameters that drive the key and value types have to be passed in at once (e.g., only the `create(...)` methods would use generics).

rhauch commented:
@guozhangwang, rebased this PR. The Jenkins build seemed to have failed on `core` -- is there a way I can resubmit it?

rhauch commented:
@guozhangwang, I just added a commit that replaces all of the static factory methods for the different kinds of key-value stores with a simple builder mechanism. Although a fluent API can be used for some parameters, most still have to be passed in to the final 'create(...)' method -- this keeps the API and the implementation very simple. (We could break up more of the parameters into separate fluent-style methods, but this very quickly makes the builder very complicated. See comments above for detail why.)

rhauch commented:
Once again, the Jenkins test failures seem irrelevant.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/SlidingWindowDef.java:
```diff
@@ -186,7 +185,7 @@ public void flush() {
             ByteArraySerializer byteArraySerializer = new ByteArraySerializer();
 
-            RecordCollector collector = ((ProcessorContextImpl) context).recordCollector();
+            RecordCollector collector = ((RecordCollector.Supplier) context).recordCollector();
```
What are the benefits of requiring ProcessorContextImpl implementing the Supplier interface rather than do the casting?

I originally think there will be two categories of users of Kafka Streaming: 1) normal users will only be interfacing the public APIs, 2) some small groups of developers that add windowing def / etc themselves will need to be exposed with the internal implementations in order to code, so it is OK to ask them to know about ProcessorContextImpl. But maybe there are something that I missed here.

rhauch replied:
@guozhangwang: The primary benefit is that we can use these components _in unit tests_ using a mock `ProcessorContext`. Without this small abstraction, all unit tests would have to use a real `ProcessorContextImpl` and everything that requires.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/RocksDBKeyValueStore.java:
```diff
@@ -253,24 +260,25 @@ public int compare(byte[] left, byte[] right) {
+                    K from, K to) {
+                super(iter, serdes);
+                iter.seek(serdes.rawKey(from));
+                this.to = serdes.rawKey(to);
```
nit: rename this.to -> this.rawToKey

rhauch replied:
ack.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -0,0 +1,206 @@
+    /**
+     * The factory for creating in-memory key-value stores.
+     */
+    public static interface InMemoryFactory extends KeyValueStoreFactory {
```
Does this factory need to extend key-value store factory? It seems a general in-memory factory, not specifically for key-value stores, right? And same for OffHeapFactory.

guozhangwang replied:
And if they are not used yet, we can probably remove them for now.

rhauch replied:
@guozhangwang, I'm not sure I follow. Can you provide a sample of how the API might be used? Given that the only `StateStore` implementations are the two `KeyValueStore` implementations, it's hard to generalize about other kinds of non-key-value stores.

 this is kind of a placeholder for future methods specific to in-memory stores, such as [a method that limits the capacity](https://github.com/rhauch/kafka/commit/7678126f4d7cd16c8bc50c632fe6403077719c81#diff-ecac0ceb71a3464bb2ce8be375242b12).

rhauch replied:
@guozhangwang, they are used. See [here](https://github.com/apache/kafka/pull/255/files#diff-a94de5b2ec72d09ebac7183c31d7c906R267).

guozhangwang replied:
What I meant is that InMemoryFactory / OffHeapFactory should probably not extending from KeyValueStoreFactory since from their names they are more general factories; for example, in the future you may have a DocumentStore / etc, and an InMemoryDocumentStore implementing this interface. InMemoryFactory will provide general functionalities like capacity limits you mentioned, but not limited to key-value store.

Now since we only have key-value stores, Stores.inMemory will use in-memory key-value stores "by default"; but moving forward we may want Stores.inMemory().keyValue().create() to separate different types. And if we follow this path then I think InMemoryFactory / OffHeapFactory should not extend from a specific KeyValueStoreFactory.

rhauch replied:
@guozhangwang. Gotcha. Perhaps the different kind of stores should be first with in-memory vs off-heap at the second level, in case not all types of stores support both. So, for example:

```
Stores.keyValue().inMemory().create(...)
```

Then that brings up several other questions:

1) Should we allow users to create untyped KeyValueStore instances by allowing them to use the ProcessorContext's default serdes? That seems error prone since the default serdes are not typed.

2) So if we remove the ability to implicitly use the default serdes, that allows us to always determine the key and value types based upon parameters. Perhaps we could pass the serdes (or the classes) into the `keyValue` method, and make 'inMemory()' or 'offHeap()' much more of a minor option.

For example, we'd be able to do this:

```
    KeyValueStore<Integer, String> store1 = Stores.create("name", context)
                                                  .withKeys(Integer.class)
                                                  .withValues(String.class)
                                                  .inMemory()
                                                  .build();
```

or

```
    KeyValueStore<Integer, String> store2 = Stores.create("name", context)
                                                  .withKeys(new IntegerSerializer(), new IntegerDeserializer())
                                                  .withValues(new StringSerializer(), new StringDeserializer())
                                                  .inMemory()
                                                  .build();
```

Do you like this better than the first option I mentioned in this comment?

guozhangwang replied:
This is what I was thinking:
1. Stores.keyValue() returns KeyValueStoreFactory that extends StoreFactory.
2. KeyValueStoreFactory has its own APIs like inMemory() / offHeap() that sets the store type and returns itself as a StoreFactory. Other XXXStoreFactory may not need to support these interface functions. For example a DocumentStoreFactory may be assumed to always off-heap.
3. If we use withValue() / withKey() to set the ser-de then we may lose the typecheck safty, so I was thinking StoreFactory.create() takes no parameters as the baseline and other extended interfaces can have other create() signatures, like including the key-value ser-des.

rhauch replied:
@guozhangwang, take a look at [this alternative](https://gist.github.com/rhauch/ad9a30bb47a041f5301d), which does ensure type safety and works with IDE code completion. Requiring some definition of the key and values (rather than using the default serdes on ProcessorContext) is what makes this work. I think this addresses all of your comments. WDYT?

guozhangwang replied:
So to clarify: if we add other store types other than key-value in the future, e.g. like a DocumentStore that does not have a key field and value field, but rather document id, secondary index, document-string field, etc.. We will need to rename create() to keyValue() / document() / etc that returns different XXFactory with API functions right?

rhauch replied:
@guozhangwang, actually, I was thinking that `create(String name, ProcessorContext)` would stay the same and would return the general `StoreFactory` interface (since all `StateStore` instances probably need a name and context), but then `StoreFactory` would have different methods for different kinds of stores. Here's how `StoreFactory` might be enhanced to add some currently-hypothetical stores:

```
public interface StoreFactory {
   /* Start of proposed methods for current KeyValueStore implementations*/
   ValueFactory<String> withStringKeys();
   ValueFactory<String> withIntegerKeys();
   ValueFactory<String> withLongKeys();
   ValueFactory<String> withKeys( Serializer<K> keySer, Deserializer<K> keyDeser );

   /* Start of hypothetical concepts that could take a variety of forms */

   // document storage ...
   DocumentFactory forDocuments();

   // queryable index storage ...
   IndexFactory usingIndexes( String definition );

   // rdbms ...
   DatabaseFactory withDatabase(...);
} 
```

See [my concept](https://gist.github.com/rhauch/ad9a30bb47a041f5301d) for more details on what it'd look like if I changed this commit. I've already verified that this code works and is easily used.

It's a tricky compromise between adding too many levels vs keeping things separated. Right now, `StoreFactory` would contain the various starter methods, while the result of those methods return `ValueFactory`, `DocumentFactory`, `IndexFactory`, `DatabaseFactory`, or something else, and all of those interfaces would completely independent.

```
DocumentStore store3 = Stores.create("name", context)
                             .forDocuments()
                             .storedCompressed()
                             .blahBlah()
                             .build();
```

If this makes sense and seems useful, my next step would be to add a commit that changes the `Stores` class to use [this new design](https://gist.github.com/rhauch/ad9a30bb47a041f5301d).

guozhangwang replied:
My minor concern is that KeyFactory and ValueFactory may be only specific to key-value stores, not not any general stores; for example for database stores you may have some functions like

```
withSchema()
```

that defines the data types for each column but not using "withKey / Value" any more. But since it is only for future improvements let's revisit this nested mechanism later.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -0,0 +1,206 @@
+         * @return this factory instance
+         * @see #withTime(Time)
+         */
+        public KeyValueStoreFactory withSystemTime();
```
Time object is only used for recording metrics, and hence users who do not create their own store implementation (i.e. the first category of users) should not be worry about what time to use; in addition other timers other than system time are used usually for tests, so I'm wondering if we want to expose those in the public APIs for the end users?

rhauch replied:
@guozhangwang, I agree it is questionable. I can remove the `withTime(...)` and `withSystemTime()` methods from the API, and if they are needed in the future we can always add them again.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/AbstractKeyValueStoreTest.java:
```diff
@@ -0,0 +1,191 @@
+                                                                      boolean useContextSerdes);
+    
+    @Test
+    public void testIntegerKeysAndStringValues() {
```
nit: the test case names are a bit too long, could we make them:

testPutGetRange / testPutGetRangeWithDefaultSerde
testRestore / testRestoreWithDefaultSerde

We can ignore the key value types since they are not for specific purposes.

rhauch replied:
ack

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java:
```diff
@@ -0,0 +1,417 @@
+     * @param key the key for the entry
+     * @param value the value for the entry
+     */
+    public void addRestoreEntry(K key, V value) {
```
A general comment: when we flush records, we should make sure the updates' corresponding logs has also been flushed so that they can be restored. So I think we do not need this function, but just change the RecordCollector above so keep a sent record list that gets updated upon send(), and move all the items in the list into another flushed list upon flush(). And when restoring we just restore all records in the flushed list.

rhauch replied:
@guozhangwang, this is actually used to populate the log **before** creating and restoring the key value store; basically, to test that a store is restoring itself properly. It should **not** be used in normal tests that use a store (including flushing). Should I improve the JavaDoc for this method?

guozhangwang replied:
I see. That makes sense. Improving the JavaDoc would be great.

guozhangwang commented:
@rhauch sorry for the late review. I think it's better to keep this PR separate from #256 (but seems it does not contain the commits from that ticket anyways). Overall I think it is good, and I like the Stores.create() mechanism except some minor comment about withTime().

rhauch commented:
@guozhangwang, incorporated your comments, rebased upon trunk, successfully ran build locally, and updated this PR.

rhauch committed:
KAFKA-2593 Added a fluent builder API for creating key-value stores

Added a simple fluent builder API to create key-value stores that use in-memory or a local database. Other kinds of state stores can be added in the future.

rhauch commented:
@guozhangwang, I've updated the PR to use the `Stores.create(...)` approach, and the full build passed locally before I pushed. The Jenkins test failure is unrelated to this PR.

rhauch commented:
@guozhangwang: Is there anything else you need from me before this can be merged? Once it's merged, I can then rebase #256.

guozhangwang commented:
LGTM, thanks @rhauch !
